### ISSUE TITLE
`F821`: Declared only variables

### ISSUE THREAD

**Issue by eltoder** [astral-sh/ruff#21494]
### Summary

Local variable annotations make annotated variables local even if there is no direct assignment. This is useful to create a local variable without an initial value. Ruff appears to not understand this usage. For example:
```python
from typing import Callable

def make_cell() -> tuple[Callable[[], int], Callable[[int], None]]:
    x: int

    def getter() -> int:
        return x

    def setter(xx: int) -> None:
        nonlocal x
        x = xx

    return getter, setter

getter, setter = make_cell()
setter(123)
print(getter())
```
(playground link: https://play.ruff.rs/3bd90cac-2c87-48b0-9cb4-64fe2a53bf72)

This runs fine and passes mypy, however ruff produces a `F821`:
```
$ ruff check --isolated t.py
F821 Undefined name `x`
 --> t.py:7:16
  |
6 |     def getter() -> int:
7 |         return x
  |                ^
8 |
9 |     def setter(xx: int) -> None:
```

### Version

0.14.5

**Comment by MichaReiser** [comment]
Thank you. Yeah, this is a bug. A variable that is declared by using an annotated assignment without a value shouldn't raise an error.

**Comment by Ruchir28** [comment]
How should this be handled?

While looking into the code, it seems we could suppress this by checking
if the annotation-only binding is inside a function and then returning
`ReadResult::Resolved(binding_id)`. But I'm not sure if returning Resolved
is the correct behavior here, since the variable may still be unassigned.

https://github.com/astral-sh/ruff/blob/fffbe5a879af0c1b36e44b65219257bcd7e54483/crates/ruff_python_semantic/src/model.rs#L467-L485

**Comment by eltoder** [comment]
@Ruchir28 it's not any different than
```python
if False:
    name = something
```

**Comment by ntBre** [comment]
@Ruchir28, from a very quick look, that seems like a reasonable place to mark the binding as resolved. Do other things break when you do that?

I'm also curious about the `BindingKind::Nonlocal` branch in `resolve_load`. It seems like `x` should possibly be marked as resolved after the `nonlocal` statement in `setter`, but I might be misunderstanding without playing with the code.

**Comment by Ruchir28** [comment]
@ntBre No it doesn't break anything - and looks like it does resolve the issue 

and I checked the `BindingKind::Nonlocal` branch ,and from what i understand it is indeed marking x as resolved in setter function.


Although not related to the fix, after fixing this, the below code won't fire F821
but still there should be F823 / UnboundLocalError right (which it doesn't currently)?
Should this be addressed separately or in this issue?

```python
def annotated_local_function():
    x: int
    print(x)
```

**Comment by eltoder** [comment]
@Ruchir28 `nonlocal` does not appear to mark `x` as resolved:
```python
def make_cell():
    def setter():
        nonlocal x
        print(x)
```
This produces F821. I think it would make sense to mark it as resolved, but it also does not materially change anything, because a variable declared as `nonlocal` must be already visible in the current scope (#21495).

For the F823 / UnboundLocalError -- it appears to be raised only in a very specific pattern. For example, this does not raise it:
```python
def annotated_local_function():
    if False:
        x = 1
    print(x)
```
My impression is that F823 is only raised when the issue is evident from a single statement like `+=` and does not require any control flow analysis. For the local annotation neither of the statements are wrong. You need some control flow analysis to discover that there is no assignment between the declaration and the print.

I suppose you can keep a state per variable of whether it is only declared or assigned, so you will report F823 on
```python
def annotated_local_function():
    x: int
    print(x)
```
but not on
```python
def annotated_local_function():
    x: int
    print(x)
    x = 1
```
I'm not sure how useful this is, but it is probably easy to implement.

**Comment by Ruchir28** [comment]
Interesting looking at the code it looks like it's marking the nonlocal as resolved , but this is weird https://github.com/astral-sh/ruff/issues/21495 
https://github.com/astral-sh/ruff/blob/78ce17ce8f068eb9ef42e3b992aa23a880722d09/crates/ruff_python_semantic/src/model.rs#L566-L581

Also regarding F823 i am not so sure about it and it's implemenation details rn but looks like it can be reported in both the cases here even in the second one [as the load reference will be attached to the initial BindingKind::Annotation id] so it can be detected there too

**Comment by SoundDesignerToBe** [comment]
Is this the same bug or another issue? Has it been reported?

```python
def test....():
    application = Application("com.domain.app",  db=session_db)
    
    def check_window(app):
        global test_success
        windows = [
            win
            for win in Gtk.Window.list_toplevels()
            if hasattr(win, "application") and win.application == application  # noqa: F821  << RUFF complains
        ]                                                                                   
        try:
            assert isinstance(windows[-1], MainWindow), (
                "Last window in list is not MainWindow"
            )
            assert len(windows) == 1, "More than 1 windows found"
        except AssertionError as err:
            logger.error(err)
        else:
            test_success = True
            logger.critical(f"MainWindow Success: {test_success}")
        finally:
            app.quit()
            return False  # stops idle loop     ■ `return` inside `finally` blocks cause exc

    original_activate = application.activate    

    def test_activate(key):
        original_activate(key)
        GLib.idle_add(check_window, application)  # noqa: F821  # << RUFF complains

    application.activate = test_activate

    application.run((__file__, str(key)))
    assert test_success
    test_success = False
    mw = Gtk.Window.list_toplevels()[0]
    mw.destroy()
    application.quit()
    del mw, application
```

I just noticed, that in the first case there is already `app` provided which works just fine, but the second nested function `test_activate` does not have an alternative to `application` and gets it via closure. The test itself works just fine, only my pre-commit hook for ruff prevents commits without the noqa addition.

**Comment by MichaReiser** [comment]
@SoundDesignerToBe this is https://github.com/astral-sh/ruff/issues/9349
